src/Analyzer.JsonRuleEngine/Converters/ExpressionConverter.cs

```diff
@@ -45,17 +57,53 @@ public override ExpressionDefinition ReadJson(JsonReader reader, Type objectType
 
             var objectPropertyNames = jsonObject.Properties().Select(property => property.Name).ToList();
 
-            // See if a property representing a structured expression is present.  If so, parse that structured expression.
-            // TODO: Parse structured expressions
+            // Add new structuredExpressions here
+            var structuredExpressions = AllOfExpressionJsonPropertyNames;
+
+            // Verify an operator property exists, representing an Expression
+            var expressionJsonPropertyNames = LeafExpressionJsonPropertyNames;
+            expressionJsonPropertyNames.UnionWith(structuredExpressions);
+
+            ValidateExpressions(jsonObject, structuredExpressions, expressionJsonPropertyNames);
+
+            var structuredExpression = objectPropertyNames.FirstOrDefault(property => structuredExpressions.Contains(property));
+            if (!string.IsNullOrEmpty(structuredExpression))
+            {
+                if (objectPropertyNames.Contains("allOf", StringComparer.OrdinalIgnoreCase))
+                {
+                    return CreateExpressionDefinition<AllOfExpressionDefinition>(jsonObject, serializer);
+                }
 
-            // Verify an operator property exists, representing a LeafExpression
-            var leafPropertyCount = objectPropertyNames.Count(property => LeafExpressionJsonPropertyNames.Contains(property));
-            if (leafPropertyCount == 1)
+                throw new JsonException($"Expression is not supported. One of the following fields is not supported: {string.Join(", ", objectPropertyNames.ToArray())}");
+            }
+            else
             {
                 return CreateExpressionDefinition<LeafExpressionDefinition>(jsonObject, serializer);
             }
+        }
+
+        private void ValidateExpressions(JObject jsonObject, HashSet<string> structuredExpressions, HashSet<string> expressionJsonPropertyNames)
+        {
+            var objectPropertyNames = jsonObject.Properties().Select(property => property.Name).ToList();
+
+            var expressionPropertyCount = objectPropertyNames.Count(property => expressionJsonPropertyNames.Contains(property));
+            if (expressionPropertyCount == 1)
+            {
+                // See if a property representing a structured expression is present.  If so, parse that structured expression.
+                var structuredExpression = objectPropertyNames.FirstOrDefault(property => structuredExpressions.Contains(property));
+                if (!string.IsNullOrEmpty(structuredExpression))
+                {
+                    // Iterate over all expressions in structured expression and validate it
+                    foreach (var expression in jsonObject.InsensitiveToken(structuredExpression))
+                    {
+                        ValidateExpressions((JObject)expression, structuredExpressions, expressionJsonPropertyNames);
+                    }
+                }
```

Comment:
You don't need to do this.  ValidateExpressions will end up being called anyway as each Expression is parsed while creating the AllOf instance.